### PR TITLE
Ticket 749 - Measurement widget

### LINKED ISSUE #749: Measure widget logic
Measure widget logic for three buttons/widgets

### PR DESCRIPTION
**Pull request by eyosef**
- `measureContent.tsx`
    - leverage `useSelector()` to access properties in `measureContent` (Redux store). Properties manage 
        - state of the area, distance and coordinates button
        - the state of the results generated when using the widgets
    - `measureContent` of `modal.config.js` is used to maintain the content of the measureContent component
    - properties of local component state `selectedAreaUnit`, `selectedDistanceUnit` and `selectedCoordinatesUnit` manage user selected dropdown option
    - `setMeasurementUnit()` => fires when user selects a specific unit from dropdown. Conditionally fires `mapController. updateSelectedMeasureWidget()` to update a specific measure widget with the correct unit. Also maintains local component state RE: user's selected unit
    - `returnMeasurementResults()` dynamically returns results based on `activeButton` (property of redux store)
    - `setSelectedWidget()` conditionally fires `setActiveMeasureWidget()`to initialize a specific measure widget. 
    - `setOption()` conditionally sets or deselects widgets by dispatching `setActiveMeasureButton()` and `setMeasureResults()`, and firing `setSelectedWidget()` to initialize the selected widget. Is reusable.
    - `returnSelectedUnit()` returns the appropriate selected unit depending on the value of `activeButton` (property of Redux state)

Fixes https://github.com/wri/gfw-mapbuilder/issues/749

### CODE REVIEW

File: src/js/controllers/mapController.ts [hunk 1]
```diff
@@ -213,6 +230,244 @@ export class MapController {
...
+    const latitudeInteger = Math.floor(latitude);
+    const latitudeMinutes = Math.floor((latitude % 1) * 60);
+    const latitudeSeconds = Math.floor((latitudeMinutes % 1) * 60);
+
+    const longitudeInteger = Math.floor(longitude);
+    const longitudeMinutes = Math.floor((longitude % 1) * 60);
+    const longitudeSeconds = Math.floor((longitudeMinutes % 1) * 60);
+
+    const latitudeInDMS = `${latitudeInteger}°${latitudeMinutes}'${latitudeSeconds}"`;
+    const longitudeDMS = `${longitudeInteger}°${longitudeMinutes}'${longitudeSeconds}"`;
```
Comment (eyosef) [inline src/js/controllers/mapController.ts]: I'll rename either `latitudeInDMS` or `longitudeDMS` (to align the constants) during tomorrow's PR review to sync changes into 1 commit

File: src/js/store/appState/types.ts [hunk 2]
```diff
@@ -1,13 +1,37 @@
...
+  toggleButton: {
+    areaButton?: boolean;
+    distanceButton?: boolean;
+    coordinatesButton?: boolean;
+  };
+  results: {
+    areaResults?: SpecificResults;
+    distanceResults?: SpecificResults;
+    coordinateMouseClickResults?: Point | any;
+    coordinatePointerMoveResults?: any;
```
Comment (eyosef) [inline src/js/store/appState/types.ts]: update `Point`

File: src/js/controllers/mapController.ts [hunk 3]
```diff
@@ -57,8 +65,14 @@ export class MapController {
...
   _sketchVM: SketchViewModel | undefined;
   _previousSketchGraphic: any;
+  _mouseClickEventListener: EventListener | any;
+  _pointerMoveEventListener: EventListener | any;
   _printTask: PrintTask | undefined;
   _legend: Legend | undefined;
+  _selectedWidget: any; // DistanceMeasurement2D | AreaMeasurement2D | undefined;
+  // * NOTE - _selectedWidget is typed as any
+  // * because ESRI's TS types measurementLabel as a string
+  // * when AreaMeasurement2D.viewModel.measurementLabel is an object
```
Comment (eyosef) [inline src/js/controllers/mapController.ts:80]: @vaidotasp @SwampGuzzler just flagging for the PR review

File: src/js/store/appState/actions.ts [hunk 4]
```diff
@@ -6,7 +6,9 @@ import {
   SET_OPEN_LAYER_GROUP,
   SET_HIDE_WIDGET,
   AppState,
-  LeftPanel
+  LeftPanel,
+  SET_MEASURE_RESULTS,
```
Comment (eyosef) [inline src/js/store/appState/actions.ts:10]: @vaidotasp @SwampGuzzler just flagging that @SwampGuzzler and I have chatted about moving measure results into the mapview reducer. Could this be a next PR?

Comment (SwampGuzzler) [reply src/js/store/appState/actions.ts:10]: @eyosef This is out of scope for this PR; we can make an Issue afterwards!

File: src/js/store/appState/reducers.ts [hunk 5]
```diff
@@ -17,6 +19,17 @@ const initialState: AppState = {
...
+  measureContent: {
+    toggleButton: {
+      activeButton: ''
+    },
+    results: {
+      areaResults: {},
+      distanceResults: {},
+      coordinateMouseClickResults: {},
+      coordinatePointerMoveResults: {}
+    }
```
Comment (vaidotasp) [inline src/js/store/appState/reducers.ts]: measureContent: {
  activeButton: ''
  areaResults:{}
  distanceResults:{}

File: src/js/store/appState/types.ts [hunk 6]
```diff
@@ -1,13 +1,42 @@
+import Point from 'esri/geometry/Point';
```
Comment (vaidotasp) [inline src/js/store/appState/types.ts]: comment it out

File: src/js/store/appState/types.ts [hunk 7]
```diff
@@ -1,13 +1,42 @@
...
+
 export interface LeftPanel {
   tabViewVisible: boolean;
   activeTab: string;
   openLayerGroup: string;
 }
 
+interface SpecificAreaResults {
+  area?: number;
```
Comment (vaidotasp) [inline src/js/store/appState/types.ts]: see if these can be required and if nothing breaks

File: src/js/components/mapWidgets/widgetContent/measureContent.tsx [hunk 8]
```diff
@@ -8,58 +22,227 @@ interface SpecificDropDownOption {
...
-    });
+        );
+      }
+    );
   };
 
-  const setAreaOption = (): void => {
-    console.log('setAreaOption');
+  const setSelectedWidget = (optionType: string): void => {
+    switch (optionType) {
```
Comment (vaidotasp) [inline src/js/components/mapWidgets/widgetContent/measureContent.tsx]: if (optionType === 'coordinates') {
///} else {
mapController.setActiveMeasureWIdget
}

File: src/js/controllers/mapController.ts [hunk 9]
```diff
@@ -57,8 +65,14 @@ export class MapController {
   _mapview: MapView | undefined;
   _sketchVM: SketchViewModel | undefined;
   _previousSketchGraphic: any;
+  _mouseClickEventListener: EventListener | any;
+  _pointerMoveEventListener: EventListener | any;
   _printTask: PrintTask | undefined;
   _legend: Legend | undefined;
+  _selectedWidget: any; // DistanceMeasurement2D | AreaMeasurement2D | undefined;
```
Comment (vaidotasp) [inline src/js/controllers/mapController.ts:77]: Create a separate issue for TS/ESRI code cleanup! :)

File: src/js/controllers/mapController.ts [hunk 10]
```diff
@@ -402,6 +417,227 @@ export class MapController {
...
+      if (state === 'measured') {
+        store.dispatch(
+          setMeasureResults({
+            areaResults,
+            distanceResults,
+            coordinateMouseClickResults: {},
+            coordinatePointerMoveResults: {}
+          })
+        );
+      }
```
Comment (vaidotasp) [inline src/js/controllers/mapController.ts]: we can extract area/perimeter/distance from selectedWidget in here, no need for the above watcher!

File: src/js/controllers/mapController.ts [hunk 11]
```diff
@@ -402,6 +417,227 @@ export class MapController {
...
+    } else if (
+      coordinateMouseClickResults?.latitude &&
+      coordinateMouseClickResults?.longitude &&
+      isDecimal
+    ) {
+      // TODO - convert DMS to decimal
+    }
+  }
+
+  updateMeasureWidgetOnClick(): void {
```
Comment (vaidotasp) [inline src/js/controllers/mapController.ts:657]: Bug": 🐛1. Area widget, draw an area, do not double click. 2) select unit from dropdown 3) hover over the map and "click" 4) drawing disappears! . Looks like we cleared the map widget content when we did not need to.

File: src/js/controllers/mapController.ts [hunk 12]
```diff
@@ -402,6 +417,227 @@ export class MapController {
...
+      store.dispatch(
+        setMeasureResults({
+          areaResults,
+          distanceResults,
+          coordinateMouseClickResults: {},
+          coordinatePointerMoveResults: {}
+        })
+      );
+    }
+    this.updateMeasureWidgetOnClick();
```
Comment (vaidotasp) [inline src/js/controllers/mapController.ts]: fire this conditionally, only when we need to.

File: src/js/controllers/mapController.ts [hunk 13]
```diff
@@ -402,6 +417,227 @@ export class MapController {
...
+        })
+      );
+    } else if (
+      coordinateMouseClickResults?.latitude &&
+      coordinateMouseClickResults?.longitude &&
+      isDecimal
+    ) {
+      // TODO - convert DMS to decimal
+    }
+  }
```
Comment (vaidotasp) [inline src/js/controllers/mapController.ts]: comment this bad boy out! 😎

File: src/js/controllers/mapController.ts (hunk 3 above)
Comment (vaidotasp) [reply src/js/controllers/mapController.ts:80]: new issue!

File: src/js/store/appState/actions.ts (hunk 4 above)
Comment (vaidotasp) [reply src/js/store/appState/actions.ts:10]: new issue!

File: src/js/store/appState/reducers.ts (hunk 5 above)
Comment (vaidotasp) [reply src/js/store/appState/reducers.ts]: new issue!

File: src/js/controllers/mapController.ts (hunk 10 above)
Comment (vaidotasp) [reply src/js/controllers/mapController.ts]: this PR

File: src/js/controllers/mapController.ts (hunk 11 above)
Comment (vaidotasp) [reply src/js/controllers/mapController.ts:657]: this PR!

File: src/js/controllers/mapController.ts (hunk 13 above)
Comment (vaidotasp) [reply src/js/controllers/mapController.ts]: this PR!

File: src/js/store/appState/types.ts (hunk 7 above)
Comment (eyosef) [reply src/js/store/appState/types.ts]: @vaidotasp @SwampGuzzler just a quick update - making these properties required causes TS errors where the logic assigns them as empty objects. Unfortunately union typing doesn't resolve those errors, so I'll keep them optional for now

**Review by vaidotasp — APPROVED** [review]
💯

**Pull request merged**